### PR TITLE
fixed permutation of VVV vertices

### PR DESCRIPTION
**Pull request by wkotlarski**
I think there might be a sign error in `VVV` vertices.
`Vertex[{conj[VWp], VWp, VZ}]` gives
```mathematica
{
{conj[VWp[{lt1}]], VWp[{lt2}], VZ[{lt3}]}, 
{I g2 Cos[ThetaW], 
   g[lt1, lt2] (-Mom[conj[VWp[{lt1}]], lt3] + Mom[VWp[{lt2}], lt3]) + 
   g[lt2, lt3] (-Mom[VWp[{lt2}], lt1] + Mom[VZ[{lt3}], lt1]) + 
   g[lt1, lt3] (Mom[conj[VWp[{lt1}]], lt2] - Mom[VZ[{lt3}], lt2])
}
}
```
while `Vertex[{VWp, conj[VWp], VZ}]` gives
```mathematica
{
{VWp[{lt1}], conj[VWp[{lt2}]], VZ[{lt3}]}, 
{-I g2 Cos[ThetaW], 
   -g[lt1, lt2] (-Mom[conj[VWp[{lt2}]], lt3] + Mom[VWp[{lt1}], lt3]) - 
   g[lt2, lt3] (Mom[conj[VWp[{lt2}]], lt1] - Mom[VZ[{lt3}], lt1]) - 
   g[lt1, lt3] (-Mom[VWp[{lt1}], lt2] + Mom[VZ[{lt3}], lt2])
}
}
```
If you look only a the value of the coupling, not the lorentz structure, then they differ by sign. But currently, there is no `SortCp` rule for `VVV` couplings. This means, that when you ask for the value of `Vertex[{conj[VWp], VWp, VZ}]` you'll get the same as for `Vertex[{VWp, conj[VWp], VZ}]`. This PR fixes this. The only thing I'm not sure is why hasn't this been a problem before? I've added Jae-hyeon to the discussion since he's the original author of `SortCp`.

Ps. This potentially also affects @dhjjacob and @uukhas

### CODE REVIEW

**Comment by Expander** [comment]
Many thanks for checking!

I don't know why this has not been a problem before, although one can imagine many cases where it may have been an issue (for example in the MW calculation or g-2). Maybe, accidentally the signs all cancel out? Anyway, it seems that all unit tests work out, so no observable seems to be affected.

Since I'm not sure what the contracts of the `SortCp` function are, I can't give a proper review here. @jhyeon may be best to comment here.

**Comment by wkotlarski** [comment]
This should afect even as basic stuff as 1-loop masses. It's possible, that such error always come in pairs though. The only test that did fail was when Jobst compared his verices with original ones in FS. He took his vertex for
```mathematica
{TreeMasses`GetWBoson[], CXXDiagrams`LorentzConjugate[TreeMasses`GetWBoson[]], TreeMasses`GetZBoson[]}
```
and checked it against `CpconjVWpVWpVZ`. This differed by a minus sign. I reoriented the vertex to 
```mathematica
{CXXDiagrams`LorentzConjugate[TreeMasses`GetWBoson[]], TreeMasses`GetWBoson[], TreeMasses`GetZBoson[]}
```
to fix the test.

**Comment by Expander** [comment]
> It's possible, that such error always come in pairs though.

Yes, this is indeed the case. For example the W contribution to the 1-loop Z self-energy contains:
~~~
result += -0.6666666666666666*AbsSqr(CpconjVWpVWpVZ())*( ... )
~~~
I.e., a sign change would not matter here.

Concerning Jobst's work: Just for my understanding, is it sure that Jobst's results are correct? (Just want to be sure.)

**Comment by wkotlarski** [comment]
It seems so. If I compare apples with apples, so `CpconjVWpVWpVZ` with Vertex<conj<Wp>,VWp,VZ>` then they agree.

**Comment by Expander** [comment]
> It seems so. If I compare apples with apples, so `CpconjVWpVWpVZ` with `Vertex<conj<Wp>,VWp,VZ>` then they agree.

Do you mean before or after the change?

**Comment by wkotlarski** [comment]
Before the change he compared `CpconjVWpVWpVZ` with `Vertex<VWp,conj<Wp>,VZ>` and it worked, although in my opinion it shouldn't. After my change this test broke. I fixed it by comparing against `Vertex<conj<Wp>,VWp,VZ>` instead.

**Comment by Expander** [comment]
Perfect, that sounds exactly how it should be.

**Comment by wkotlarski** [comment]
@iolojz, if you have a moment, could you comment on this PR?

**Comment by jhyeon** [comment]
Sorry for the slow reponse!

> But currently, there is no `SortCp` rule for `VVV` couplings.

There _is_ a `SortCP` rule for `VVV` couplings.  For instance
```Mathematica
Vertices`SortCp @ SARAH`Cp[conj[VWm], VWm, VZ]
Vertices`SortCp @ SARAH`Cp[VWm, conj[VWm], VZ]

Out[1]= Cp[conj[VWm], VWm, VZ]
Out[2]= Cp[conj[VWm], VWm, VZ]
```
Reordering of fields in `Cp[]` is done by ``Vertices`SortFieldsInCp[]``.  It mimics `SortCoup[]` in `SARAH/Package/deriveModel.m`.  Then further operations are optionally applied to the `Cp[]` depending on the vertex type.  These operations are again for reproducing the (C++-equivalent of) code from ``SARAH`MakeSPheno[]``.

**Comment by wkotlarski** [comment]
I'm sorry @jhyeon. My mistake. I shouldn't have said 'identity'. What I wanted to say is that if falls under generic rule - sort arguments of ```SARAH`Cp``` without adding extra signs. Maybe I misunderstood how should I use this function. I want to limit the number of vertices that have to written by not using vertices that are related by permuation of fields. I understood that this is what `SortCp` does. If I need vertex `Vertex[{X, Z, Y}]` it should give me it as a function of canonically ordered vertex, lets say `Vertex[{X, Y, Z}]`. As you noted
```mathematica
Vertices`SortCp @ SARAH`Cp[conj[VWm], VWm, VZ] === Vertices`SortCp @ SARAH`Cp[VWm, conj[VWm], VZ]
```
but in SARAH
```mathematica
Vertext[{conj[VWm], VWm, VZ}] != Vertices[{VWm, conj[VWm], VZ}]
```

**Comment by jhyeon** [comment]
I wrote `SortCp[]` only to reproduce the SPheno subroutines from SARAH that calculate 1-loop self energies.  The file `FS/Output/MSSM/EWSB/One-Loop/SelfEnergy.m` generated by SARAH contains the self energy for instance of `VWm`
```Mathematica
{VWm, ...
-conj[Cp[conj[VWm], VZ, VWm]] Cp[conj[VWm], VZ, VWm] (A0[Mass2[VWm]] + A0[Mass2[VZ]] + ...)
+ ...}
```
The corresponding lines in `Output/MSSM/EWSB/SPheno/LoopMasses_MSSM.f90` look like
```FORTRAN
Subroutine Pi1LoopVWm()
[snip]
!------------------------ 
! VZ, VWm 
!------------------------ 
sumI = 0._dp 
coup1 = cplcVWmVWmVZ
coup2 = Conjg(coup1) 
    SumI = -VVVloop(p2,MVZ2,MVWm2)*coup1*coup2 
res = res +1._dp* SumI  
[snip]
res = oo16pi2*res 
End Subroutine Pi1LoopVWm 
```
where `VVVloop()` is equal to the parenthesized loop factor in the above Mathematica expression and the coupling `cplcVWmVWmVZ` is calculated in `Couplings_MSSM.f90`:
```FORTRAN
Subroutine CouplingcVWmVWmVZL(g2,TW,res)
[snip]
res = 0._dp
res = res-(g2*Cos(TW))
[snip]
End Subroutine CouplingcVWmVWmVZL
```
As suggested by the order of particles appearing in the subroutine name, this coupling is equal to `-I` times the coefficient for the "sorted" list of fields:
```Mathematica
-I SARAH`Vertex[{conj[VWm], VZ, VWm}][[2, 1]] (* unsorted *)
-I SARAH`Vertex[{conj[VWm], VWm, VZ}][[2, 1]] (* sorted *)

Out[3]= g2 Cos[ThetaW]
Out[4]= -g2 Cos[ThetaW]
```
In short, `SortCp[]` brings `Cp[]`s in `SelfEnergy.m` closer to the couplings that SPheno calculates.

> What I wanted to say is that if falls under generic rule - sort arguments of `` SARAH`Cp `` without adding extra signs. Maybe I misunderstood how should I use this function. I want to limit the number of vertices that have to written by not using vertices that are related by permuation of fields. I understood that this is what `SortCp` does.

Perhaps you are right.  `SortCp[]` may have been working so far since all the extra minus signs it may have missed cancelled out.  If your PR does not break (or even fixes) self energies, then it would make a lot of sense to me.

**Comment by wkotlarski** [comment]
@Expander, @pathron, can I rebase and merge this branch now?

**Comment by Expander** [comment]
@wkotlarski Yes, please feel free to rebase-and-merge it.
